addons/misra.py

```diff
@@ -740,8 +747,35 @@ def getForLoopCounterVariables(forToken):
                     vars_modified.add(tn.variable)
                 elif tn.previous and tn.previous.str in ('++', '--'):
                     vars_modified.add(tn.variable)
-        if cur_clause == 1 and tn.isAssignmentOp and tn.astOperand1.variable:
-            vars_initialized.add(tn.astOperand1.variable)
+        if cur_clause == 1 and tn.isAssignmentOp:
+            if tn.astOperand1.variable:
+                vars_initialized.add(tn.astOperand1.variable)
+            elif tn.astOperand1.str == "." and tn.astOperand1.astOperand2 and tn.astOperand1.astOperand2.variable:
+                vars_initialized.add(tn.astOperand1.astOperand2.variable)
+        if cur_clause == 1 and tn.function:
+            function_args_in_init = getArguments(tn.next);
+            function_args = tn.function.argument
+            function_scope = get_function_scope(cfg, tn.function)
+            if function_scope:
+                for key,args in function_args.items():
+                    if args.nameToken and args.nameToken.valueType and args.nameToken.valueType.pointer > 0:
+                        function_body_start = function_scope.bodyStart
+                        function_body_end = function_scope.bodyEnd
+                        while function_body_start != function_body_end:
```

Comment:
there should be some utility function that checks if variable is changed in a block of code.

In core cppcheck we have `isVariableChanged` it is declared in the astutils.h. It might make sense to implement similar utility function in misra.py or cppcheckdata.py.